File: impl/ocaml/src/design_system.ml
(* The design system: theme, registries, memoization caches, and the set of
   known-invalid candidates (SPEC §4.1.9). *)

open Candidate

type t = {
  theme : Theme.t;
  utilities : Utilities.t;
  variants : Variants.t;
  invalid_candidates : (string, unit) Hashtbl.t;
  (* Marks every generated declaration !important. *)
  mutable important : bool;
  candidate_cache : (string, candidate list) Hashtbl.t;
  variant_cache : (string, variant) Hashtbl.t;
  variant_missing : (string, unit) Hashtbl.t;
  mutable variant_order : (string, int) Hashtbl.t option;
}

let create theme =
  { theme; utilities = Utilities.create (); variants = Variants.create (); invalid_candidates = Hashtbl.create 64; important = false;
    candidate_cache = Hashtbl.create 1024; variant_cache = Hashtbl.create 64; variant_missing = Hashtbl.create 64; variant_order = None }

(* Parses a variant, memoized so equal inputs share one value. *)
let rec parse_variant ds raw =
  match Hashtbl.find_opt ds.variant_cache raw with
  | Some v -> Some v
  | None -> (
      if Hashtbl.mem ds.variant_missing raw then None
      else
        match Candidate.parse_variant raw (context ds) with
        | None ->
            Hashtbl.replace ds.variant_missing raw ();
            None
        | Some v ->
            Hashtbl.replace ds.variant_cache raw v;
            ds.variant_order <- None;
            Some v)

(* The parser context backed by this design system. *)
and context ds : Candidate.context =
  { theme_prefix = (fun () -> ds.theme.prefix);
    has_utility = Utilities.has ds.utilities;
    has_variant = Variants.has ds.variants;
    variant_kind_of = Variants.kind ds.variants;
    compounds_with = Variants.compounds_with ds.variants;
    parse_variant = parse_variant ds }

(* Creates a design system with the built-in variants and utilities. *)
let build theme =
  let ds = create theme in
  Builtin_variants.register ds.variants theme;
  Builtin_utilities.register ds.utilities theme;
  ds

(* Returns every interpretation of a raw class name, memoized. *)
let parse_candidate ds raw =
  match Hashtbl.find_opt ds.candidate_cache raw with
  | Some cached -> cached
  | None ->
      let parsed = Candidate.parse_candidate raw (context ds) in
      Hashtbl.replace ds.candidate_cache raw parsed;
      parsed

(* Sorts every parsed variant and assigns an index that increases each time
   the comparison reports a difference (SPEC §9.4). Keyed by the variant text. *)
let variant_order ds =
  match ds.variant_order with
  | Some order -> order
  | None ->
      let parsed = Hashtbl.fold (fun _ v acc -> v :: acc) ds.variant_cache [] in
      let parsed = List.stable_sort (fun a b -> compare a.vraw b.vraw) parsed in
      let parsed = List.stable_sort (Variants.compare ds.variants) parsed in
      let order = Hashtbl.create (List.length parsed) in
      let index = ref 0 and previous = ref None in
      List.iter
        (fun v ->
          (match !previous with Some p when Variants.compare ds.variants p v <> 0 -> incr index | _ -> ());
          Hashtbl.replace order v.vraw !index;
          previous := Some v)
        parsed;
      ds.variant_order <- Some order;
      order

let order_of ds (v : variant) = match Hashtbl.find_opt (variant_order ds) v.vraw with Some i -> i | None -> 0

(* Drops memoized candidates. *)
let clear_candidate_cache ds = Hashtbl.reset ds.candidate_cache

File: impl/go/design_system.go
package twill

import "sort"

// DesignSystem owns the theme, the registries, memoization caches, and the
// set of known-invalid candidates (SPEC §4.1.9).
type DesignSystem struct {
	Theme             *Theme
	Utilities         *Utilities
	Variants          *Variants
	InvalidCandidates map[string]bool
	// Important marks every generated declaration !important.
	Important bool

	candidateCache map[string][]*Candidate
	variantCache   map[string]*Variant
	variantMissing map[string]bool
	compileCache   map[*Candidate]map[int][]CompiledRule
	variantOrder   map[*Variant]int
}

// NewDesignSystem creates an empty design system.
func NewDesignSystem(theme *Theme) *DesignSystem {
	return &DesignSystem{
		Theme:             theme,
		Utilities:         NewUtilities(),
		Variants:          NewVariants(),
		InvalidCandidates: map[string]bool{},
		candidateCache:    map[string][]*Candidate{},
		variantCache:      map[string]*Variant{},
		variantMissing:    map[string]bool{},
		compileCache:      map[*Candidate]map[int][]CompiledRule{},
	}
}

// BuildDesignSystem creates a design system with the built-in variants and utilities.
func BuildDesignSystem(theme *Theme) *DesignSystem {
	ds := NewDesignSystem(theme)
	RegisterBuiltinVariants(ds.Variants, theme)
	RegisterBuiltinUtilities(ds.Utilities, theme)
	return ds
}

// ThemePrefix implements ParserContext.
func (ds *DesignSystem) ThemePrefix() string { return ds.Theme.Prefix }

// HasUtility implements ParserContext.
func (ds *DesignSystem) HasUtility(name string, kind UtilityKind) bool {
	return ds.Utilities.Has(name, kind)
}

// HasVariant implements ParserContext.
func (ds *DesignSystem) HasVariant(name string) bool { return ds.Variants.Has(name) }

// VariantKindOf implements ParserContext.
func (ds *DesignSystem) VariantKindOf(name string) (VariantKind, bool) { return ds.Variants.Kind(name) }

// CompoundsWith implements ParserContext.
func (ds *DesignSystem) CompoundsWith(parent string, child *Variant) bool {
	return ds.Variants.CompoundsWith(parent, child)
}

// ParseCandidate returns every interpretation of a raw class name, memoized.
func (ds *DesignSystem) ParseCandidate(raw string) []*Candidate {
	if cached, ok := ds.candidateCache[raw]; ok {
		return cached
	}
	parsed := ParseCandidate(raw, ds)
	ds.candidateCache[raw] = parsed
	return parsed
}

// ParseVariant parses a variant, memoized so equal inputs share one object.
func (ds *DesignSystem) ParseVariant(raw string) *Variant {
	if v, ok := ds.variantCache[raw]; ok {
		return v
	}
	if ds.variantMissing[raw] {
		return nil
	}
	v := ParseVariant(raw, ds)
	if v == nil {
		ds.variantMissing[raw] = true
		return nil
	}
	ds.variantCache[raw] = v
	ds.variantOrder = nil
	return v
}

// VariantOrder sorts every parsed variant and assigns an index that
// increases each time Compare reports a difference (SPEC §9.4).
func (ds *DesignSystem) VariantOrder() map[*Variant]int {
	if ds.variantOrder != nil {
		return ds.variantOrder
	}
	parsed := make([]*Variant, 0, len(ds.variantCache))
	for _, v := range ds.variantCache {
		parsed = append(parsed, v)
	}
	sort.SliceStable(parsed, func(i, j int) bool { return ds.Variants.Compare(parsed[i], parsed[j]) < 0 })
	order := map[*Variant]int{}
	index := 0
	var previous *Variant
	for _, v := range parsed {
		if previous != nil && ds.Variants.Compare(previous, v) != 0 {
			index++
		}
		order[v] = index
		previous = v
	}
	ds.variantOrder = order
	return order
}

// ClearCandidateCache drops memoized candidates and compiled results.
func (ds *DesignSystem) ClearCandidateCache() {
	ds.candidateCache = map[string][]*Candidate{}
	ds.compileCache = map[*Candidate]map[int][]CompiledRule{}
}
